lib/tree.rb

```diff
@@ -1,64 +1,176 @@
+    #       if !current.left.nil?
+    #         current = current.left
+    #       else
+    #         current.left = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     else
+    #       if !current.right.nil?
+    #         current = current.right
+    #       else
+    #         current.right = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     end
+    #   end
+    # end
+    
+    @root = add_helper(key, value, @root)
+    return @root
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def add_helper(key, value, current)
+    if current.nil?
+      return TreeNode.new(key, value)
+    end
+    
+    if key <= current.key
+      current.left = add_helper(key, value, current.left)
+    else
+      current.right = add_helper(key, value, current.right)
+    end
+    
+    return current
+  end
+  
+  # Time Complexity: O(n) if the BST starts to resembles a linked list
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    current = @root
+    
+    until current.nil?
+      if current.key == key
+        return current.value
+      elsif key < current.key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
+    
+    return nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n) We are passing by reference?
   def inorder
-    raise NotImplementedError
+    # return an array
+    inorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def inorder_helper(current, list)
+    return list if current.nil?
+    
+    # we're passing by reference, so we don't have to set list = inorder_helper calls
+    # left - root - right
+    inorder_helper(current.left, list)
+    list << { key: current.key, value: current.value }
+    inorder_helper(current.right, list)
+    
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    preorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def preorder_helper(current, list)
+    return list if current.nil?
+    
+    # root - left - right
+    list << { key: current.key, value: current.value }
+    
+    preorder_helper(current.left, list)
+    preorder_helper(current.right, list)
+    
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    postorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def postorder_helper(current, list)
+    return list if current.nil?
+    
+    # left-right-root
+    postorder_helper(current.left, list)
+    postorder_helper(current.right, list)
+    list << { key: current.key, value: current.value }
+    
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n) for the count variables?
   def height
```

Comment:
👍  Nice work!

However the space complexity is O(n) for an unbalanced tree (based off the call stack), and O(log n) for a balanced tree.